librad/src/git/local/transport.rs

```diff
@@ -0,0 +1,426 @@
+            let mut tracing_enabled = false;
+            tracing::error!(
+                traced = {
+                    tracing_enabled = true;
+                    tracing_enabled
+                },
+                "{}",
+                msg
+            );
+            if !tracing_enabled {
+                eprintln!("{}", msg)
+            }
+        });
+    }
+}
+
+/// Connect the forked service's stdio to any [`Stdio`].
+///
+/// The service is either `git-upload-pack` or `git-receive-pack` depending on
+/// the [`Service`] passed to [`LocalTransport::connect`].
+pub struct Localio {
+    pub child_stdin: Stdio,
+    pub child_stdout: Stdio,
+}
+
+impl Localio {
+    /// Arrange for pipes between parent and child.
+    pub fn piped() -> Self {
+        Self {
+            child_stdin: Stdio::piped(),
+            child_stdout: Stdio::piped(),
+        }
+    }
+
+    /// Inherit stdio from the parent.
+    pub fn inherit() -> Self {
+        Self {
+            child_stdin: Stdio::inherit(),
+            child_stdout: Stdio::inherit(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct LocalTransport {
+    storage: Arc<Mutex<Storage<WithSigner>>>,
+}
+
+impl LocalTransport {
+    pub fn new(settings: Settings) -> Result<Self, Error> {
+        let storage = Storage::open(&settings.paths)?.with_signer(settings.signer)?;
+        Ok(LocalTransport {
+            storage: Arc::new(Mutex::new(storage)),
+        })
+    }
+
+    pub fn stream(
+        &mut self,
+        url: LocalUrl,
+        service: Service,
+        stdio: Localio,
+    ) -> Result<LocalStream, Error> {
+        let mut child = self.connect(url, service, Mode::Stateless, stdio)?;
+
+        let stdin = child.process.stdin.take().unwrap();
+        let stdout = child.process.stdout.take().unwrap();
+
+        // Spawn a thread to `wait(2)` on the child process.
+        //
+        // Note that we have no straightforward and portable way to forward
+        // signals to the child process in Rust. When `LocalStream` is dropped,
+        // `ChildStdin` and `ChildStdin` will close the underlying file
+        // descriptors, which should cause `SIGPIPE` to be delivered to the
+        // child. Unless `git` is somehow misbehaving, this should guarantee
+        // that this thread is not leaked.
+        thread::spawn(move || {
+            child
+                .wait()
+                .map_err(|e| tracing::error!("Error waiting on child process: {}", e))
+        });
+
+        let header = match service {
+            Service::UploadPackLs => Some(UPLOAD_PACK_HEADER.to_vec()),
+            Service::ReceivePackLs => Some(RECEIVE_PACK_HEADER.to_vec()),
+            _ => None,
+        };
+
+        Ok(LocalStream {
+            read: LocalRead {
+                header,
+                inner: stdout,
+            },
+            write: stdin,
+        })
+    }
+
+    pub fn connect(
+        &mut self,
+        url: LocalUrl,
```

Comment:
Should we use the `Into<RadUrn>` pattern here to allow us to pass in `RadUrn` as well? Or was there a specific reason we need `LocalUrl` here?

Reply:
Sure you can, but is there a specific reason to make it more polymorphic?

The remote helper gets invoked with `<name of remote> <url of remote>`, eg. `git remote-rad origin rad://deadf4laf31`. The URL is a `LocalUrl`. Similarly, the transport `action` is also called with a `LocalUrl`.